### PR TITLE
Remove `DiagCtxt` API duplication

### PR DESCRIPTION
**Pull request by nnethercote**
`DiagCtxt` defines the internal API for creating and emitting diagnostics: methods like `struct_err`, `struct_span_warn`, `note`, `create_fatal`, `emit_bug`. There are over 50 methods.

Some of these methods are then duplicated across several other types: `Session`, `ParseSess`, `Parser`, `ExtCtxt`, and `MirBorrowckCtxt`. `Session` duplicates the most, though half the ones it does are unused. Each duplicated method just calls forward to the corresponding method in `DiagCtxt`. So this duplication exists to (in the best case) shorten chains like `ecx.tcx.sess.parse_sess.dcx.emit_err()` to `ecx.emit_err()`.

This API duplication is ugly and has been bugging me for a while. And it's inconsistent: there's no real logic about which methods are duplicated, and the use of `#[rustc_lint_diagnostic]` and `#[track_caller]` attributes vary across the duplicates.

This PR removes the duplicated API methods and makes all diagnostic creation and emission go through `DiagCtxt`. It also adds `dcx` getter methods to several types to shorten chains. This approach scales *much* better than API duplication; indeed, the PR adds `dcx()` to numerous types that didn't have API duplication: `TyCtxt`, `LoweringCtxt`, `ConstCx`, `FnCtxt`, `TypeErrCtxt`, `InferCtxt`, `CrateLoader`, `CheckAttrVisitor`, and `Resolver`. These result in a lot of changes from `foo.tcx.sess.emit_err()` to `foo.dcx().emit_err()`. (You could do this with more types, but it gets into diminishing returns territory for types that don't emit many diagnostics.)

After all these changes, some call sites are more verbose, some are less verbose, and many are the same. The total number of lines is reduced, mostly because of the removed API duplication. And consistency is increased, because calls to `emit_err` and friends are always preceded with `.dcx()` or `.dcx`.

r? @compiler-errors

### CODE REVIEW

**Comment by nnethercote** [comment]
Here's a table showing the full API and the existing duplication.
```
DC: DiagCtxt
S:  Session
PS: ParseSess
P:  Parser
EC: ExtCtxt
MB: MirBorrowckCtxt

func                              DC S PS P EC MB
----                              ---------------
struct_diagnostic                 DC     

struct_bug                        DC     

struct_fatal                      DC S PS
struct_span_fatal                 DC S   
struct_span_fatal_with_code       DC S     

struct_almost_fatal               DC       

struct_err                        DC S PS  
struct_span_err                   DC S    P EC
struct_err_with_code              DC S     
struct_span_err_with_code         DC S         MB  

struct_err_lint                   DC       

struct_warn                       DC S PS 
struct_span_warn                  DC S    
struct_warn_with_code             DC S    
struct_span_warn_with_code        DC S   

struct_warn_with_expectation      DC S      
struct_span_warn_with_expectation DC S     

struct_note                       DC S    
struct_span_note                  DC     

struct_help                       DC     

struct_allow                      DC S   
struct_span_allow                 DC S   

struct_expect                     DC S   

----

bug                               DC        EC  
span_bug                          DC      P EC 
span_bug_no_panic                 DC         

span_delayed_bug                  DC S       
good_path_delayed_bug             DC S       

fatal                             DC S       
span_fatal                        DC S       
span_fatal_with_code              DC S       

err                               DC S       
span_err                          DC S      EC    
span_err_with_code                DC S       

warn                              DC S      
span_warn                         DC S      
span_warn_with_code               DC S      

note                              DC S     
span_note                         DC S     

----

create_err                        DC S PS   EC  
create_warning                    DC S PS    
create_almost_fatal               DC        
create_fatal                      DC S PS   
create_bug                        DC        
create_note                       DC S PS   

----

emit_err                          DC S PS   EC 
emit_warning                      DC S PS    
emit_almost_fatal                 DC         
emit_fatal                        DC S PS   
emit_bug                          DC        
emit_note                         DC S PS   
```

**Comment by nnethercote** [comment]
And here's a table showing all the possible chaining combinations.
```
ParseSess
- old
  - parse_sess.emit_err()
  - parse_sess.dcx.emit_err()
- new
  - parse_sess.dcx.emit_err()

Session
- old
  - sess.emit_err(e)
  - sess.parse_sess.emit_err(e)
  - sess.parse_sess.dcx.emit_err(e)
- new
  - sess.dcx().emit_err()
  - sess.parse_sess.dcx.emit_err(e)

Parser [note: `sess` here is a `ParseSess`]
- old
  - parser.emit_err()
  - parser.sess.emit_err()
  - parser.sess.dcx.emit_err()
- new
  - parser.dcx().emit_err()
  - parser.sess.dcx.emit_err()

TyCtxt (no current API duplication)
- old
  - tcx.sess.emit_err()
  - tcx.sess.parse_sess.emit_err(e)
  - tcx.sess.parse_sess.dcx.emit_err(e)
- new
  - tcx.dcx().emit_err()
  - tcx.sess.dcx().emit_err()
  - tcx.sess.parse_sess.dcx.emit_err()

ExtCtxt
- old
  - ecx.emit_err()
  - ecx.tcx.sess.emit_err()
  - ecx.tcx.sess.parse_sess.emit_err()
  - ecx.tcx.sess.parse_sess.dcx.emit_err()
- new
  - ecx.dcx().emit_err()
  - ecx.tcx.dcx().emit_err()
  - ecx.tcx.sess.dcx().emit_err()
  - ecx.tcx.sess.parse_sess.dcx.emit_err()

Other TyCtxt wrappers without a dcx() getter
- old
  - foo.tcx.sess.emit_err()
  - foo.tcx.sess.parse_sess.emit_err()
  - foo.tcx.sess.parse_sess.dcx.emit_err()
- new (if you don't add a dcx() getter)
  - foo.tcx.dcx().emit_err()
  - foo.tcx.sess.dcx().emit_err()
  - foo.tcx.sess.parse_sess.dcx.emit_err()
- new (if you add a dcx() getter)
  - foo.dcx().emit_err()
  - foo.tcx.dcx().emit_err()
  - foo.tcx.sess.dcx().emit_err()
  - foo.tcx.sess.parse_sess.dcx.emit_err()
```
The shortest chains are always preferred, of course.

File: compiler/rustc_const_eval/src/interpret/eval_context.rs [hunk 1]
```diff
@@ -473,9 +473,9 @@ impl<'mir, 'tcx: 'mir, M: Machine<'mir, 'tcx>> InterpCx<'mir, 'tcx, M> {
         backtrace.print_backtrace();
         // FIXME(fee1-dead), HACK: we want to use the error as title therefore we can just extract the
         // label and arguments from the InterpError.
-        let dcx = self.tcx.sess.dcx();
+        let dcx = self.tcx.dcx();
         #[allow(rustc::untranslatable_diagnostic)]
-        let mut diag = self.tcx.sess.struct_allow("");
+        let mut diag = self.tcx.dcx().struct_allow("");
```
Comment (RalfJung) [inline compiler/rustc_const_eval/src/interpret/eval_context.rs:478]:
```suggestion
        let mut diag = dcx.struct_allow("");
```

Comment (nnethercote) [reply compiler/rustc_const_eval/src/interpret/eval_context.rs]: Fixed, thanks.

**Comment by nnethercote** [comment]
Not expected to affect perf, but let's check:

@bors try @rust-timer queue

**Comment by nnethercote** [comment]
I fixed the conflicts.

**Comment by nnethercote** [comment]
I figure I won't fix the conflicts again until/unless this gets r+, because it's so conflict-prone. The conflicts won't affect reviewing in any significant way.

**Comment by compiler-errors** [comment]
Yeah, sorry, this one's quite large. No need to rebase until I say `r=me`.

**Review by compiler-errors — APPROVED** [review]

**Comment by compiler-errors** [comment]
r=me, sorry for the delay

**Comment by nnethercote** [comment]
Thanks for the review. I did a rebase the other day anyway so I think this is mergeable.

@bors r=compiler-errors rollup=never p=1

High priority because it's prone to conflicts.

**Pull request merged**